src/Raven.Server/Documents/Subscriptions/AbstractSubscriptionStorage.cs

```diff
@@ -17,28 +17,45 @@
 
 namespace Raven.Server.Documents.Subscriptions;
 
-public abstract class AbstractSubscriptionStorage<TState> : ILowMemoryHandler, IDisposable
-    where TState : AbstractSubscriptionConnectionsState
+public abstract class AbstractSubscriptionStorage
 {
-    protected readonly ConcurrentDictionary<long, TState> _subscriptions = new();
-    public ConcurrentDictionary<long, TState> Subscriptions => _subscriptions;
     protected readonly ServerStore _serverStore;
     protected string _databaseName;
     protected readonly SemaphoreSlim _concurrentConnectionsSemiSemaphore;
     protected Logger _logger;
 
+    protected abstract string GetNodeFromState(SubscriptionState taskStatus);
+    protected abstract DatabaseTopology GetTopology(ClusterOperationContext context);
+    public abstract bool DropSingleSubscriptionConnection(long subscriptionId, string workerId, SubscriptionException ex);
+
     protected AbstractSubscriptionStorage(ServerStore serverStore, int maxNumberOfConcurrentConnections)
     {
         _serverStore = serverStore;
         _concurrentConnectionsSemiSemaphore = new SemaphoreSlim(maxNumberOfConcurrentConnections);
-        LowMemoryNotification.Instance.RegisterLowMemoryHandler(this);
     }
 
-    protected abstract void DropSubscriptionConnections(TState state, SubscriptionException ex);
-    protected abstract void SetConnectionException(TState state, SubscriptionException ex);
-    protected abstract string GetSubscriptionResponsibleNode(DatabaseRecord databaseRecord, SubscriptionState taskStatus);
-    protected abstract bool SubscriptionChangeVectorHasChanges(TState state, SubscriptionState taskStatus);
-    public abstract bool DropSingleSubscriptionConnection(long subscriptionId, string workerId, SubscriptionException ex);
+    public string GetSubscriptionResponsibleNode(ClusterOperationContext context, SubscriptionState taskStatus)
+    {
+        var topology = GetTopology(context);
+        var tag = GetNodeFromState(taskStatus);
+        var lastFunc = GetLastResponsibleNode(_serverStore.LicenseManager.HasHighlyAvailableTasks(), topology, tag);
+        return topology.WhoseTaskIsIt(_serverStore.Engine.CurrentState, taskStatus, lastFunc);
+    }
+
+    public static string GetSubscriptionResponsibleNodeForProgress(RawDatabaseRecord record, string shardName, SubscriptionState taskStatus, bool hasHighlyAvailableTasks)
```

Comment:
we should avoid code like this, this is why you have abstraction layer

Reply:
This method is used in the cluster state machine, which have no abstraction.